### ISSUE TITLE
how to do PCA visualization

### ISSUE THREAD

**Issue by Mythili-kannan** [facebookresearch/vjepa2#148]
Thanks for the awesome work, I'm very much interested in visualizing the PCA features as show in the paper, i tried few techniques, but couldnt replicate as clean as in paper, can someone help me on that

**Comment by gajeshladhar** [comment]
Hi, 

you can use this code snippet : 

```
TUBELET_SIZE = 2 
PATCH_SIZE   = 16 


with torch.no_grad():
    y = encoder([x])[0] # x.shape : (N,C,T,H,W)

from sklearn.decomposition import PCA
y_np = y.cpu().numpy().reshape(-1, 1024)  # shape: (n_tokens, 1024)
pca = PCA(n_components=3)
pca.fit(y_np)
y_pca = pca.transform(y_np)  # shape: (n_tokens, 3)

T2 = s2.shape[2] // TUBELET_SIZE

n_tokens = y.shape[1]
spatial_tokens = n_tokens // T2                    # H2 * W2
H2 = s2.shape[-2] // PATCH_SIZE                   # naive guess
W2 = spatial_tokens // H2                         # infer W2 from actual token count

y_pca = y_pca.reshape(T2,H2, W2, 3)                  # (H2, W2, 3)
```

**Comment by Mythili-kannan** [comment]
T, H, W = 32, 24, 24
model_hub = backbones.vjepa2_1_vit_giant_384(pretrained=True)[0]
model_hub.cuda().eval()
video_transform = build_pt_video_transform(img_size=384)
vr = VideoReader("/root/sample_video.mp4")
frame_idx = np.linspace(0, len(vr) - 1, 64).astype(int)

video = vr.get_batch(frame_idx).asnumpy()
x = video_transform(video).cuda().unsqueeze(0)
with torch.no_grad():
    tokens = model_hub(x)
tokens = tokens[0].reshape(T, H,W, -1).cpu().numpy()
flat_tokens = tokens.reshape(-1, D)
flat_tokens = torch.tensor(flat_tokens)
flat_tokens = F.normalize(flat_tokens, dim=1).numpy()
pca = PCA(n_components=3)
pca_features = pca.fit_transform(flat_tokens)
# # normalize to [0, 1]
pca_min = pca_features.min(axis=0)
pca_max = pca_features.max(axis=0)
pca_features = (pca_features - pca_min) / (pca_max - pca_min + 1e-8)
pca_features = pca_features.reshape(T, H, W, 3)

output_frames = []
for t in range(T):
    rgb_map = pca_features[t]
    rgb_map = cv2.resize(
        rgb_map,
        (480, 270),
        interpolation=cv2.INTER_NEAREST
    )
    rgb_map = (rgb_map * 255).astype(np.uint8)
    output_frames.append(cv2.cvtColor(rgb_map, cv2.COLOR_RGB2BGR))
    save_path = os.path.join(out_dir, f"frame_{t}.png")
    cv2.imwrite(save_path, cv2.cvtColor(rgb_map, cv2.COLOR_RGB2BGR))


i used this snippet, the maps looks okayish, but not as same as in the paper
